### PR TITLE
defensive check

### PR DESCRIPTION
**Pull request by tbo47**
to address https://github.com/tbo47/dagre-es/pull/34#discussion_r1770577444

@aloisklink

### CODE REVIEW

File: src/dagre/position/bk.js [hunk 1]
```diff
@@ -71,7 +71,7 @@ function findType1Conflicts(g, layering) {
     return layer;
   }
 
-  layering.reduce(visitLayer);
+  if (Array.isArray(layering) && layering.length > 0) layering.reduce(visitLayer);
```
Comment (aloisklink) [inline src/dagre/position/bk.js:74]: I think it's fine to skip `Array.isArray()`.

I had a look at the code and `layering` comes from:

https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/dagre/util.js#L135-L145

which always returns an array (although, sometimes an empty array).

**Review by aloisklink — CHANGES_REQUESTED** [review]
Thanks for making this fix! I didn't have the time to do it last night!

I think we also need to add the defensive check here: https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/dagre/debug.js#L22-L25

---

I also think the `Array.isArray()` is not needed, so we could change the code to the following:

```diff
diff --git a/src/dagre/debug.js b/src/dagre/debug.js
index 434f957..12573ef 100644
--- a/src/dagre/debug.js
+++ b/src/dagre/debug.js
@@ -19,10 +19,12 @@ function debugOrdering(g) {
   layerMatrix.forEach((layer, i) => {
     var layerV = 'layer' + i;
     h.setNode(layerV, { rank: 'same' });
-    layer.reduce((u, v) => {
-      h.setEdge(u, v, { style: 'invis' });
-      return v;
-    });
+    if (layer.length > 0) {
+      layer.reduce((u, v) => {
+        h.setEdge(u, v, { style: 'invis' });
+        return v;
+      });
+    }
   });

   return h;
diff --git a/src/dagre/position/bk.js b/src/dagre/position/bk.js
index df57531..66f792d 100644
--- a/src/dagre/position/bk.js
+++ b/src/dagre/position/bk.js
@@ -71,7 +71,9 @@ function findType1Conflicts(g, layering) {
     return layer;
   }

-  layering.reduce(visitLayer);
+  if (layering.length > 0) {
+    layering.reduce(visitLayer);
+  }
   return conflicts;
 }

@@ -114,7 +116,9 @@ function findType2Conflicts(g, layering) {
     return south;
   }

-  layering.reduce(visitLayer);
+  if (layering.length > 0) {
+    layering.reduce(visitLayer);
+  }
   return conflicts;
 }
```

**Comment by tbo47** [comment]
Thanks. Do you think we need to do th

**Comment by aloisklink** [comment]
> Thanks. Do you think we need to do the defensive check elsewhere?

Ah, we do! Sorry, I think I had gone halfway through checking these last night, and forgot to do the rest this morning!

- https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/dagre/rank/network-simplex.js#L183-L189
- https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/graphlib/graph.js#L315-L322

I believe all of the other `.reduce()` calls set an `initialValue`, so if the array is empty, they're return that instead of throwing an error.

**Comment by tbo47** [comment]
> > Thanks. Do you think we need to do the defensive check elsewhere?
> 
> Ah, we do! Sorry, I think I had gone halfway through checking these last night, and forgot to do the rest this morning!
> 
> * https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/dagre/rank/network-simplex.js#L183-L189
> * https://github.com/tbo47/dagre-es/blob/e03371f2ee986cd30ac5b3229e18816980e8fb7d/src/graphlib/graph.js#L315-L322
> 
> I believe all of the other `.reduce()` calls set an `initialValue`, so if the array is empty, they're return that instead of throwing an error.

done

File: src/dagre/rank/network-simplex.js [hunk 2]
```diff
@@ -173,20 +173,22 @@ function enterEdge(t, g, edge) {
...
+    return candidates.reduce((acc, edge) => {
+      if (slack(g, edge) < slack(g, acc)) {
+        return edge;
+      }
 
-    return acc;
-  });
+      return acc;
+    });
+  }
```
Comment (aloisklink) [inline src/dagre/rank/network-simplex.js:191]: In the future, we could also add the following, to make it a bit more explicit that `enterEdge` can return `undefined`, but if ESLint isn't giving us a warning/error, it's fine for now.

```suggestion
  } else {
    return undefined;
  }
```

**Review by aloisklink — APPROVED** [review]
Looks good to me! Thanks for making those changes!

**Pull request merged**